### PR TITLE
apps/cmp.c: Fix double free on OSSL_CMP_CTX_set1_p10CSR() failure

### LINKED ISSUE #14910: [BUG] A double free bug in openssl/apps/cmp.c
**File:** openssl/apps/cmp.c
**Bug Function:** setup_request_ctx
**Version:** Git-master(2021-4-18)

**Description:** In function `setup_request_ctx`,  `csr` is obtained via `load_csr_autofmt()` at line 1,583.
At line 1,587, `csr` is freed by `X509_REQ_free(csr)` at line 1,587 in the first time and `goto oom`.
But in the `oom` branch, `csr` is freed by `X509_REQ_free(csr)`  again at line 1,694, which is a double free bug.

**Code Description:**
```
static int setup_request_ctx(OSSL_CMP_CTX *ctx, ENGINE *engine)
{
    ...
    if (opt_csr != NULL) {
        if (opt_cmd == CMP_GENM) {
            CMP_warn("-csr option is ignored for command 'genm'");
        } else {
1583:       csr = load_csr_autofmt(opt_csr, "PKCS#10 CSR");
            if (csr == NULL)
                return 0;
            if (!OSSL_CMP_CTX_set1_p10CSR(ctx, csr)) {
1587:           X509_REQ_free(csr);   // csr is Freed Here !
                goto oom;
            }
        }
    }
    ....
 oom:
         CMP_err("out of memory");
 exts_err:
         sk_X509_EXTENSION_pop_free(exts, X509_EXTENSION_free);
1694:    X509_REQ_free(csr);   // csr is Freed again !
         return 0;
}


```

### PR DESCRIPTION
**Pull request by DDvO**
Fixes #14910
Also slightly improve further error handling of setup_request_ctx().

### CODE REVIEW

File: apps/cmp.c [hunk 1]
```diff
@@ -1607,15 +1604,13 @@ static int setup_request_ctx(OSSL_CMP_CTX *ctx, ENGINE *engine)
             goto exts_err;
         }
         OSSL_CMP_CTX_set0_reqExtensions(ctx, exts);
-        exts = NULL;
```
Comment (t8m) [inline apps/cmp.c:1610]: Is this correct?? set0 means the OSSL_CMP_CTX takes ownership, or not?

Comment (t8m) [reply apps/cmp.c:1610]: Ah, I see we do not jump to the err labels from here in the code. I wonder if this does not make the code fragile.

Comment (DDvO) [reply apps/cmp.c:1610]: Yes, `ctx` takes ownership.
This removal is part of the simplification of the error handling I did along with the fix.
So far, a later `goto oom` had been possible, which I removed, 
so from the above change onward there cannot be an extra free of `exts`.

Comment (t8m) [reply apps/cmp.c:1610]: Maybe add a comment that there must be no goto oom afterwards?

Comment (DDvO) [reply apps/cmp.c:1610]: Yeah, or even better if I re-add the two NULL assignments, just to be on the safe side?

Comment (t8m) [reply apps/cmp.c:1610]: I wonder if static analyzers or gcc won't start screaming about value assigned but not used.

Comment (DDvO) [reply apps/cmp.c:1610]: Good point. So added a warning comment.

**Review by t8m — APPROVED** [review]

File: apps/cmp.c (hunk 1 above)
Comment (DDvO) [reply apps/cmp.c:1610]: BTW, I wonder why Coverity did not find/report that potential (but unlikely) double free?

Comment (t8m) [reply apps/cmp.c:1610]: Static analysis is never perfect.

**Comment by DDvO** [comment]
Merged - thanks for the exchange

**Pull request closed without merging**